pymeasure/instruments/toptica/adapters.py

```diff
@@ -0,0 +1,129 @@
+#
+# Copyright (c) 2013-2020 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import re
+import time
+
+from pymeasure.adapters import VISAAdapter
+
+
+class TopticaAdapter(VISAAdapter):
+    """ Adapter class for connecting to Toptica Console via a serial
+    connection.
+
+    :param port: pyvisa resource name of the instrument
+    :param baud_rate: communication speed, defaults to 115200
+    :param kwargs: Any valid key-word argument for VISAAdapter
+    """
+    # compiled regular expression for finding numerical values in reply strings
+    _reg_value = re.compile(r"\w+\s+=\s+(\w+)")
+
+    def __init__(self, port, baud_rate, **kwargs):
+        kwargs.setdefault('preprocess_reply', self.extract_value)
+        kwargs.setdefault('read_termination', '\r\n')
+        kwargs.setdefault('write_termination', '\r\n')
+        super().__init__(port, **kwargs)
+        # hack to set baud_rate since VISAAdapter filters it out when spcified
+        # as keyword argument. issue #334
+        self.connection.baud_rate = baud_rate
+        # configure communication mode
+        super().write('echo off')
+        super().write('prom off')
+        time.sleep(0.04)
+        # clear the initial messages from the controller
+        self.flush()
+        self.ask('talk usual')
+
+    def extract_value(self, reply):
+        """ preprocess_reply function which tries to extract <value> from 'name
+        = <value> [unit]'. If <value> can not be identified the orignal string
+        is returned.
+
+        :param reply: reply string
+        :returns: string with only the numerical value, or the original string
+        """
+        r = self._reg_value.search(reply)
+        if r:
+            return r.groups()[0]
+        else:
+            return reply
+
+    def check_acknowledgement(self, reply):
+        """ checks if reply is '[OK]', otherwise a ValueError is raised and the
```

Comment:
`[OK]` is even returned after an error or invalid command. I think the `[OK]` has to be understood more as an indicator that the device is ready to receive more commands. In case an invalid command is sent an error code is returned before the `[OK]`.

e.g. ">" indicates send things, "<" received lines

```
> 'sh invalid\r\n'
< '\r\n'
< '%SYS-W-027, command unknown'
< '[OK]\r\n'
```